master_key_rotation_for_disconnected_minions.md

```diff
@@ -0,0 +1,81 @@
+
+
+# Motivation
+[motivation]: #motivation
+
+If a minion is offline when the master changes its key, there is no way to get it after the fact. The minion will not be able to reconnect to the master after the master has changed its key. This is a problem e.g. with end user desktops/laptops that aren't ALWAYS connected to the master.
+
+The goal is a key management that allows for
+- Key rotation/replacement.
+- Distribution of the replacement key while the current key is valid or at install.
+- Keeping the private replacement key offline to safeguard against a compromised master.
+
+
+Related issues:
+- https://github.com/saltstack/salt/issues/59149
+- https://github.com/saltstack/salt/issues/59342
+
+
+# Design
+[design]: #detailed-design
+
+- Definition of any new terminology
+  - Current key (k1): the key that is currently in use.
+  - Replacement key (k2): a key which is unused as long k1 works.
+  - Rotation: replacing k1 with k2.
+- Examples of how the feature is used
+  - Create replacement key k2.
+  - Distribute k2 to all minions, which could require to wait for offline minions,
+  - After complete distribution of k2, the master can rotate keys from k1 to k2.
+  - When a minion reconnects with k1 and fails, he reconnects with k2.
+- Additionial command
+  - `salt * key_util.set_replacement_key public_key=/full/path/to/file`
+    - The return of `key_util.set_replacement_key` should inform if a former replacement key was replaced and if they differed.
+- How do you plan to test it?
+  - Distribute k2 to minions m1 and m2.
+  - Disconnect m2.
+  - Rotate key on master.
+  - Verify m1 rotates key.
+  - Reconnect m2.
+  - Verify m2 rotates key.
+
+
+## Alternatives
+[alternatives]: #alternatives
+
+What other designs have been considered?
+  - Expiring key: a key that is only valid until its expire time. This seems to add no benefit.
+
+What is the impact of not doing this?
+- No key rotation when minions disconnect
+- No Key rotation API
+
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Why should we *not* do this? Please consider:
+
+- Implementation cost, both in term of code size and complexity
+  - No change on the master required.
+  - A minion must additionally store the replacement key.
+  - The key rotation action is essentially moving a file.
+- Integration of this feature with other existing and planned features
+  - The advice [ROTATE A MASTER KEY](https://docs.saltproject.io/en/latest/topics/hardening.html#rotate-a-master-key) contains a security hole because the minion must blindly trust the host that responds as `master`.
+  - There is no API to rotate keys, only a [rekey script](https://github.com/dwoz/salt-rekey/) that requires all minions to be connected at the same time.
+- Cost of migrating existing Salt setups (is it a breaking change?)
+  - No migration costs, not a breaking change.
+- Documentation (would Salt documentation need to be re-organized or altered?)
+  - Only added.
```

Comment:
There are also security considerations that should be discussed and addressed: how do we audit the keys & range changes? Currently it's basically static, but we're adding the ability for an attacker to potentially add (a) new key(s). This is especially important for the `public_key=/full/path/to/file` - we need to ensure that the file is only writable(readable??) by root or whoever Salt is.

If the key is readable by non-Salt users that means that it's pretty trivial to MITM salt, and if it's writable well... same problems but even worse :upside_down_face:

I think generally I like the idea of this SEP, but this SEP should be more detailed than a typical SEP, especially around the security considerations. Given the area this touches we should get as many eyes on it as possible, including the security WG at *minimum*. We should get some VMware security review as well.

Reply:
@waynew wrote

> we need to ensure that the file is only writable(readable??) by root or whoever Salt is.

You need this already today and it is ensured by verify.py today.

Reply:
@waynew wrote

> we're adding the ability for an attacker to - potentially add (a) new key(s)

That's not true. What you call "ability" requires access to the API.
Attackers who can access the Salt API can add a key today. The current API is powerful enough.


However, the API under discussion makes it easier to rotate keys. This increases security, as described.

Reply:
IIUC (it's been a while since I've been in that code) - currently master key rotation is an out of band process, and requires the Salt master to restart even to enable a reload of the key. And like the highlander - there can be only one.

This isn't necessarily more or less secure (as that depends on your objectives). But currently *salt* doesn't really support changing out the key.

And if the key changes then minions will complain loudly.

But that's not the case if we change things to allow multiple keys. I'm not saying that this is a gaping security hole, but I'm saying that we should be mindful of the *possibilities*.